portfolio/src/main/java/com/google/sps/servlets/PhotoFormHandlerServlet.java

```diff
@@ -0,0 +1,120 @@
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+import com.google.gson.Gson;
+import com.google.sps.data.Post;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.*;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/photo-form-handler")
+public class PhotoFormHandlerServlet extends HttpServlet {
+    private static final String MESSAGE_PARAMETER = "message";
+    private static final String IMAGE_PARAMETER = "image";
+    private static final String POST_PARAMETER = "Post";
+    private static final String URL_PARAMETER = "url";
+    private static final String DISPLAY_PARAMETER = "display";
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String message = request.getParameter(MESSAGE_PARAMETER);
+        String imageUrl = getUploadedFileUrl(request, IMAGE_PARAMETER);
+        String display = request.getParameter(DISPLAY_PARAMETER);
+
+        Entity post = new Entity(POST_PARAMETER);
+        post.setProperty(MESSAGE_PARAMETER, message);
+        post.setProperty(URL_PARAMETER, imageUrl);
+        post.setProperty(DISPLAY_PARAMETER, display);
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        datastore.put(post);
+        response.sendRedirect("/photo.html");
+    }
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        Gson gson = new Gson();
+        
+        List<Post> posts = getPosts();
+
+        response.setContentType("application/json;");
+        response.getWriter().println(gson.toJson(posts));
+    }
+
+    private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+        BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+        Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+        List<BlobKey> blobKeys = blobs.get(IMAGE_PARAMETER);
+
+        // User submitted form without selecting a file, so we can't get a URL. (devserver)
+        if(blobKeys == null || blobKeys.isEmpty()) {
+            return null;
```

Comment:
You don't have to do this in this PR but it'd be good to add some sort of error handling so that the user gets a message about them needing to select a file.